kore/src/Kore/Step/Simplification/TermLike.hs

```diff
@@ -201,101 +225,80 @@ simplifyInternal term predicate =
-                    Ceil.simplify predicate =<< simplifyChildren ceilF
-                EqualsF equalsF ->
-                    Equals.simplify predicate =<< simplifyChildren equalsF
-                ExistsF exists ->
-                    let fresh =
-                            Lens.over
-                                Binding.existsBinder
-                                refreshBinder
-                                exists
-                    in  Exists.simplify =<< simplifyChildren fresh
-                IffF iffF ->
-                    Iff.simplify =<< simplifyChildren iffF
-                ImpliesF impliesF ->
-                    Implies.simplify =<< simplifyChildren impliesF
-                InF inF ->
-                    In.simplify predicate =<< simplifyChildren inF
-                NotF notF ->
-                    Not.simplify =<< simplifyChildren notF
-                --
-                BottomF bottomF ->
-                    Bottom.simplify <$> simplifyChildren bottomF
-                BuiltinF builtinF ->
-                    Builtin.simplify <$> simplifyChildren builtinF
-                DomainValueF domainValueF ->
-                    DomainValue.simplify <$> simplifyChildren domainValueF
-                FloorF floorF -> Floor.simplify <$> simplifyChildren floorF
-                ForallF forall ->
-                    let fresh =
-                            Lens.over
-                                Binding.forallBinder
-                                refreshBinder
-                                forall
-                    in  Forall.simplify <$> simplifyChildren fresh
-                InhabitantF inhF ->
-                    Inhabitant.simplify <$> simplifyChildren inhF
-                MuF mu ->
-                    let fresh = Lens.over Binding.muBinder refreshBinder mu
-                    in  Mu.simplify <$> simplifyChildren fresh
-                NuF nu ->
-                    let fresh = Lens.over Binding.nuBinder refreshBinder nu
-                    in  Nu.simplify <$> simplifyChildren fresh
-                -- TODO(virgil): Move next up through patterns.
-                NextF nextF -> Next.simplify <$> simplifyChildren nextF
-                OrF orF -> Or.simplify <$> simplifyChildren orF
-                RewritesF rewritesF ->
-                    Rewrites.simplify <$> simplifyChildren rewritesF
-                TopF topF -> Top.simplify <$> simplifyChildren topF
-                --
-                StringLiteralF stringLiteralF ->
-                    return $ StringLiteral.simplify (getConst stringLiteralF)
-                VariableF variableF ->
-                    return $ Variable.simplify (getConst variableF)
+          = assertTermNotPredicate $ tracer termLike $ do
+            termOr <- descendAndSimplify termLike
+            termPredicateList <- BranchT.gather $ do
+                termOrElement <- BranchT.scatter termOr
+                simplified <- simplifyCondition termOrElement
```

Comment:
How is this going to work with `simplifyCondition` calling back into the term simplifier?

Reply:
Not that well, it seems that I missed some stuff when moving code around. Thank you for pointing it out. I rewrote it, in the new version it should work like this (siw = simplifyInternalWorker, sc = simplifyCondition):
```
1. siw t1 computes t2 and p2, then calls sc p2
2. sc p2 calls siw p2
3. siw p2 computes t3 and p3. t3 should be top. If p2 == p3, i.e. this was fully simplified, mark it as such and continue. Otherwise call sc p2, go to step 2
```